### PR TITLE
Add Timeshift to VuPlus Addon

### PR DESCRIPTION
**Pull request by phunkyfish**
Hi, I've added Timeshift to the Addon basing it on the DVBViewer implementation.

It contains the config menu items and supports Timeshift Off, Timeshift On Pause and Timeshift On Playback.

There is an issue that @ksooo let me know about that disables timeshift on pause: https://github.com/xbmc/xbmc/pull/14389

### CODE REVIEW

File: pvr.vuplus/resources/language/resource.language.en_us/strings.po [hunk 1]
```diff
@@ -131,3 +131,35 @@ msgstr "Disconnected from '%s'"
...
+msgid "On playback"
+msgstr "On playback"
+
+msgctxt "#30065"
+msgid "On pause"
+msgstr "On pause"
+
+msgctxt "#30514"
+msgid "Timeshift buffer path does not exist"
+msgstr ""
```
Comment (ksooo) [inline pvr.vuplus/resources/language/resource.language.en_us/strings.po]: Please revert the changes to this file. The source language for this project is en_GB and only that strings.po needs to modified by the deveopers. All other languages are provided via Transifex and those changes would overwrite what you have changed here.

File: src/LocalizedString.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+private:
+  LocalizedString() = delete;
+  LocalizedString(const LocalizedString&) = delete;
+  LocalizedString &operator =(const LocalizedString&) = delete;
+
+  std::string m_localizedString;
+};
+
+} //namespace dvbviewer
```
Comment (ksooo) [inline src/LocalizedString.h]: dvbviewer => vuplus

File: src/VuData.cpp [hunk 3]
```diff
@@ -355,7 +404,7 @@ void  *Vu::Process()
 bool Vu::LoadChannels() 
 {    
   bool bOk = false;
-
+
```
Comment (ksooo) [inline src/VuData.cpp]: a space splipped in here

File: src/VuData.cpp [hunk 4]
```diff
@@ -364,7 +413,7 @@ bool Vu::LoadChannels()
     if (LoadChannels(myGroup.strServiceReference, myGroup.strGroupName))
       bOk = true;
   }
-
+
```
Comment (ksooo) [inline src/VuData.cpp]: two blanks slipped in here

File: src/VuData.h [hunk 5]
```diff
@@ -45,8 +47,12 @@ typedef enum VU_UPDATE_STATE
     VU_UPDATE_STATE_NEW
 } VU_UPDATE_STATE;
 
-struct VuEPGEntry 
+class VuEPGEntry
```
Comment (ksooo) [inline src/VuData.h]: IMO it does not make sense to change structs to classes if the only change is to introduce default contructors and to keep all data members public. Please revert those changes.

File: src/client.cpp [hunk 6]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+/* live stream functions */
+bool OpenLiveStream(const PVR_CHANNEL &channel)
+{
+  if (!VuData || !VuData->IsConnected())
+    return false;
+
+  if (!VuData->OpenLiveStream(channel))
+    return false;
+
+  //const Settings &settings = DvbData->GetSettings();
```
Comment (ksooo) [inline src/client.cpp]: Another leftover from stuff you copied from dvbviewer. Pleaes this line as it is already commented out.

File: src/client.cpp [hunk 7]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+{
+  return (strReader) ? strReader->IsRealTime() : false;
+}
+
+bool CanPauseStream(void)
+{
+  if (!VuData)
+    return false;
+
+  //const Settings &settings = DvbData->GetSettings();
```
Comment (ksooo) [inline src/client.cpp]: same here. please delete this line

File: src/client.cpp [hunk 8]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+  return false;
+}
+
+bool CanSeekStream(void)
+{
+  if (!VuData)
+    return false;
+
+  // pause button seems to check CanSeekStream() too
+  //return (strReader && strReader->IsTimeshifting());
```
Comment (ksooo) [inline src/client.cpp]: sidenote: checking IsTimeshifting would even be wrong here.

File: src/client.cpp [hunk 9]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+bool IsTimeshifting(void)
+{
+  return (strReader && strReader->IsTimeshifting());
+}
+
+PVR_ERROR GetStreamTimes(PVR_STREAM_TIMES *times)
+{
+    
+
+  if (!times)
```
Comment (ksooo) [inline src/client.cpp]: mind deleting the empty two lines above?

File: src/client.cpp [hunk 10]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+  return PVR_ERROR_NOT_IMPLEMENTED;
+}
+
+void PauseStream(bool paused)
+{
+  if (!VuData)
+    return;
+
+  /* start timeshift on pause */
+  //const Settings &settings = DvbData->GetSettings();
```
Comment (ksooo) [inline src/client.cpp]: dvbviewer leftover again

File: src/client.cpp [hunk 11]
```diff
@@ -533,9 +547,139 @@ PVR_ERROR GetChannelGroupMembers(ADDON_HANDLE handle, const PVR_CHANNEL_GROUP &g
...
+  if (paused && g_iEnableTimeshift == TIMESHIFT_ON_PAUSE
+      && strReader && !strReader->IsTimeshifting()
+      && XBMC->DirectoryExists(g_strTimeshiftBufferPath.c_str()))
+  {
+    strReader = new TimeshiftBuffer(strReader, g_strTimeshiftBufferPath ,g_iReadTimeout);
+    (void)strReader->Start();
```
Comment (ksooo) [inline src/client.cpp]: what's this leading  `(void)` good for?

File: src/client.h [hunk 12]
```diff
@@ -53,6 +58,20 @@ extern bool                      g_bOnlinePicons;
...
+extern int                       g_iReadTimeout;
 extern std::string               g_strChannelDataPath;
 extern ADDON::CHelper_libXBMC_addon *   XBMC;
 extern CHelper_libXBMC_pvr *     PVR;
+
+/*!
+ * @brief PVR macros for string exchange
+ */
+#define PVR_STRCPY(dest, source) do { strncpy(dest, source, sizeof(dest)-1); dest[sizeof(dest)-1] = '\0'; } while(0)
+#define PVR_STRCLR(dest) memset(dest, 0, sizeof(dest))
```
Comment (ksooo) [inline src/client.h]: This is not used.

File: src/client.h [hunk 13]
```diff
@@ -53,6 +58,20 @@ extern bool                      g_bOnlinePicons;
...
+extern std::string               g_strTimeshiftBufferPath;
+extern int                       g_iReadTimeout;
 extern std::string               g_strChannelDataPath;
 extern ADDON::CHelper_libXBMC_addon *   XBMC;
 extern CHelper_libXBMC_pvr *     PVR;
+
+/*!
+ * @brief PVR macros for string exchange
+ */
+#define PVR_STRCPY(dest, source) do { strncpy(dest, source, sizeof(dest)-1); dest[sizeof(dest)-1] = '\0'; } while(0)
```
Comment (ksooo) [inline src/client.h]: This is only used in `Signalstatus`. I would not introduce a macro if is only used two times. But this is only my personal POV.

File: src/client.h [hunk 14]
```diff
@@ -53,6 +58,20 @@ extern bool                      g_bOnlinePicons;
...
+#define PVR_STRCLR(dest) memset(dest, 0, sizeof(dest))
+
+namespace vuplus
+{
+
+}
```
Comment (ksooo) [inline src/client.h]: ??? what are the above lnes good for?

**Comment by ksooo** [comment]
Although I cannot runtime test your changes due to lacking equipment I suppose your changes should work just fine as the essential parts of your PR are copied from pvr.dvbviewer, which I know working.

**Comment by phunkyfish** [comment]
Ok, I've made the changes you suggested. I left the PVR_STRCPY as I'm not sure of the equivalent inline code. Copied from DVB Viewer.

**Comment by phunkyfish** [comment]
I might look at GetTimerTypes and Recordings next to implement the later API functionality. Have to do a bit of study first to figure out how that works.

File: src/client.cpp (hunk 11 above)
Comment (slo617) [reply src/client.cpp]: Casting a function to void is a hint that the return value is not used and that the programmer is aware of it. It might also suppress a warning if lint is used.
TimeshiftBuffer->Start() returns a bool, which is ignored here.

**Comment by phunkyfish** [comment]
Ok, so I’ll put this back. Let me know if you want it out again.

**Comment by phunkyfish** [comment]
That should be everything now added a fix for recording duration. Is there any action to be taken on my side to merge this or is that done by someone else?

File: pvr.vuplus/resources/language/resource.language.en_gb/strings.po [hunk 15]
```diff
@@ -142,3 +142,35 @@ msgstr ""
...
+msgid "On playback"
+msgstr "On playback"
+
+msgctxt "#30065"
+msgid "On pause"
+msgstr "On pause"
+
+msgctxt "#30514"
+msgid "Timeshift buffer path does not exist"
+msgstr "Timeshift buffer path does not exist"
```
Comment (ksooo) [inline pvr.vuplus/resources/language/resource.language.en_gb/strings.po]: could you please add a line break.

File: pvr.vuplus/addon.xml.in [hunk 16]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <addon
   id="pvr.vuplus"
-  version="3.7.3"
+  version="3.7.4"
```
Comment (ksooo) [inline pvr.vuplus/addon.xml.in]: As you have introduced a new feature, an update of the minor version is more appropriate => 3.8.0

File: pvr.vuplus/changelog.txt [hunk 17]
```diff
@@ -1,3 +1,10 @@
+v3.7.4
```
Comment (ksooo) [inline pvr.vuplus/changelog.txt]: 3.8.0

File: pvr.vuplus/changelog.txt [hunk 18]
```diff
@@ -1,3 +1,10 @@
...
+- Fixed: timeshift support for Live TV in v18
```
Comment (ksooo) [inline pvr.vuplus/changelog.txt]: This not a fix, but a new feature. I would phrase it simply "Added: timeshift support"

File: pvr.vuplus/changelog.txt [hunk 19]
```diff
@@ -1,3 +1,10 @@
+v3.7.4
+- Fixed: timeshift support for Live TV in v18 
+- Fixed recording duration bug and cleaned up split string code
+
+v3.7.3
+- Translation update from Transifex
+
```
Comment (ksooo) [inline pvr.vuplus/changelog.txt]: could you add also v3.7.2 (was also a translation update)?

**Comment by phunkyfish** [comment]
Apologies if I'm doing this slightly arse ways. I have all of two days practice with git. Should I use the rebase editor to combine the commits into more sensible functional commits?

I don't want to make it more messy than it is already!

**Comment by phunkyfish** [comment]
I guess you learn by doing. I see now that I should have left the PR just for Timeshift and used other PRs for the other changes.

**Comment by phunkyfish** [comment]
I’ve attempted to combine commits locally without success. I’ll need some instruction on how to do this cleanly. 

Also a question. When you comment on the PR how do I make those change to the PR without committing post review? Or are those commits acceptable just nothing unrelated.

Do you do a merge and squash into master or just merge the single commits?

**Comment by ksooo** [comment]
I create a merge commit which preserves the single commits of the PR.

**Comment by ksooo** [comment]
What I do to prevent chaos when doing changes requested by a reviewer is to first think about which of these changes belong to which existing commit, then I create a new commit per existing commit, containing the requested changes that belong to this commit. After that i do a `git rebase -i` and `fixup` the new commits after the existing commits.

**Comment by ksooo** [comment]
For your case I would do a `git reset master` (assuming your branch is based on master) and restart from scratch, by first saving a diff, then removing everything that is not related to the timeshift feature, then doing the timeshift commit. After that, re-add  changes for next commit -- signalstate impl maybe --, commit. And so on, until all is readded. 

This is for sure no fun, but there is no better way I'm afraid.

Next time you should try not to mess up things in the first place.

**Comment by phunkyfish** [comment]
Ok, I can do that. So do I create a new PR when I'm done and discard this one? Or even better a few different PRs containing the separate pieces of work?

**Comment by ksooo** [comment]
No need for a new PR or even several new PRs. It's all just git 'magic'. Just get your things done locally and then do a `git push origin -ff <your-branch-name>`

**Comment by phunkyfish** [comment]
Just like you said, it's git magic. Thanks for your help. That should be everything the way it should be.

**Comment by phunkyfish** [comment]
You're very welcome
